### DISCUSSION TITLE
Apache Hop Server  systemd

### DISCUSSION THREAD

**Discussion by ravamo** [apache/hop#4540] [Q&A] [answered]
Hi Community ,
I have two questions and I wonder if anyone can help.

- I have Apache Hop running as a systemd service, but I want to avoid having to use ExecStop=/bin/kill -TERM $MAINPID when stopping it. I’ve seen something like this on the website: ExecStop=/usr/bin/curl -X GET http://localhost:8080/hop-server/stop?password=$HOP_SERVER_ADMIN_PASSWORD. What’s the best way to stop the server properly?
 version 1
 
 ``` shell
 [Unit]
Description=Apache Hop Server
After=network.target

[Service]
Type=simple
User=hopuser
Group=hopuser
WorkingDirectory=/ruta/a/apache-hop
ExecStart=/ruta/a/apache-hop/hop-server.sh
ExecStop=/bin/kill -TERM $MAINPID
Restart=on-failure
RestartSec=5

# Limita el tamaño máximo del log
StandardOutput=syslog
StandardError=syslog
SyslogIdentifier=hop

[Install]
WantedBy=multi-user.target
```
 
 version 2
 
 ```shell
[Unit]
Description=Apache Hop Server
After=network.target

[Service]
Type=simple
User=hopuser
Group=hopuser
WorkingDirectory=/ruta/a/apache-hop
Environment="HOP_SERVER_ADMIN_PASSWORD=tu_password"
ExecStart=/ruta/a/apache-hop/hop-server.sh
ExecStop=/usr/bin/curl -X GET http://localhost:8080/hop-server/stop?password=$HOP_SERVER_ADMIN_PASSWORD
Restart=on-failure
RestartSec=5

# Limita el tamaño máximo del log
StandardOutput=syslog
StandardError=syslog
SyslogIdentifier=hop

[Install]
WantedBy=multi-user.target
```

- Also, has anyone created an Ansible playbook for the installation? If not, I can create one, but I just want to avoid duplicating efforts.
best

**Comment by kranskydog** [comment]
from https://hop.apache.org/manual/latest/hop-server/index.html
-k, --kill              Stop the running hopServer server.  This is only
                            allowed when using the hostname/port form of the
                            command. Use the -s and -u options to authenticate
                            
example: hop-server.bat 127.0.0.1 8080 --kill --userName cluster --password cluster

**Reply by hansva** [reply]
That one has been removed as it was not working as expected

**Reply by ravamo** [reply]
@hansva , In that case, what’s the best option? Running the KILL command is pretty abrupt—isn't there a way to shut it down more gracefully?

**Comment by ravamo** [comment]
roger roger ,
in that case the the right systemd is :
```shell
[Unit]
Description=Apache Hop Server
After=network.target

[Service]
Type=simple
User=hopuser
Group=hopuser
WorkingDirectory=/ruta/a/apache-hop
ExecStart=/ruta/a/apache-hop/hop-server.sh 0.0.0.0 8080
ExecStop=/ruta/a/apache-hop/hop-server.sh 0.0.0.0 8080 8182 --kill --userName cluster --password cluster
Restart=on-failure
RestartSec=5

#Max log
StandardOutput=syslog
StandardError=syslog
SyslogIdentifier=hop

[Install]
WantedBy=multi-user.target
```
In that case, is it a good idea to add the ShutdownPort (8182), or is it not needed as a best practice?

**Reply by hansva** [reply]
Currently there is no clean way to shut down a Hop Server, @nadment is taking a look if we can add something to make it possible

**Comment by ravamo** [comment]
Regarding this, is there any way to send both server logs and execution logs to a specific file, like /var/logs/apache_hop.log? I saw this [link](https://hop.apache.org/manual/latest/hop-server/index.html#_enable_detailed_server_logging), but it’s not 100% clear to me

**Reply by bamaer** [reply, marked as answer]
there isn't anything out of the box that would let send server and execution logs to the same log file, but I guess you could make it work with a combination of server configuration and [pipeline log](https://hop.apache.org//manual/latest/metadata-types/pipeline-log.html) and [workflow log](https://hop.apache.org//manual/latest/metadata-types/workflow-log.html).

**Reply by ravamo** [reply]
OK, but on the website I do have the execution logs. Isn't there somewhere to send that information to a log? I'll still check what you've mentioned

**Reply by bamaer** [reply]
by "website" you mean the Hop Server web UI? that information is not persisted anywhere at the moment, it won't survive a server restart.
